site/source/_posts/2017/11/17-Dojo-2-Beta-4/Dojo-2-Beta-4.md

```diff
@@ -0,0 +1,263 @@
+    ]);
+  }
+}
+```
+
+Can be written as this using when using Custom Elements:
+
+```html
+<demo-menu id="menu">
+  <demo-menu-item id="a" title="Menu Item A"></demo-menu-item>
+  <demo-menu-item id="b" title="Menu Item B"></demo-menu-item>
+  <demo-menu-item id="c" title="Menu Item C" selected></demo-menu-item>
+</demo-menu>
+<script>
+  a.data = { foo: 'bar' };
+  menu.addEventListener('selected', function(event) {
+    const detail = event.detail;
+    console.log(detail);
+  });
+</script>
+```
+
+We have an [live example](https://dojo.github.io/examples/custom-element-menu/) of this, along with the [source code on GitHub](https://github.com/dojo/examples/tree/master/custom-element-menu).
+
+## Other widget system improvements
+
+There were other changes to the widgeting system which enable some features and capabilities we think can make widgets more robust:
+
+* The way event listeners behave on virtual DOM nodes has changed, and we now allow the listeners to change between renders.  Previously this would have errored.
+* There is a lot more flexibility about what the _root_ of a `Projector` can be, which includes being able to return a widget node, or an array of nodes.  Also, previously there were challenges with changing the root of the projector after the first render, which is now allowed.
+* Several internal performance and memory improvements to the virtual DOM and widgeting system.
+
+## New widgets
+
+We have introduced two new widgets in Beta 4 and are considering more before the release of Dojo 2.  Really we want to focus on trying to hone our patterns for widgets to provide examples for others to build widgets, but do know that especially for rapid prototyping it is helpful to have _out-of-the-box_ widgets.
+
+We have introduced a new accordion pane widget (`@dojo/widgets/accordionpane/AccordionPane`):
+
+<img src="./accordion.png" alt="The accordion pane widget" height="465" width="351" />
+
+As well as we have introduced a tooltip widget (`@dojo/widgets/tooltip/Tooltip`):
+
+<img src="./tooltip.png" alt="The tooltip widget" height="104" width="415" />
+
+## Other widget improvements
+
+We have also been trying to standardise our conventions in widgets in Beta 4, though we will likely have more changes for Beta 5.  In particular we have been rationalising our styles, making sure that themeable styles and the structural styles are clearly delineated, making it easier to theme existing widgets.  We have also been looking at the ways to better customise and extend the out of the box widgets.
+
+## Stores re-architecture
+
+The current API in Dojo 1 for managing state focused on a CRUD type API.  Originally in Dojo 2 we directly adapted that API as part of _stores_.  Recently though, the concept of state containers has taken front and centre and we knew we wanted to create a solution that aligned to that architecture.
+
+Instead of CRUD type API, it focuses on three main concepts:
```

Comment:
Instead of CRUD type API, it focuses on three main concepts: -> Instead of a CRUD type API, `stores` focuses on three main concepts: